src/lib/instrument-listen.js

```diff
@@ -5,57 +5,66 @@ const http = require('http');
 const https = require('https');
 const denodeify = require('denodeify');
 const path = require('path');
+
 const fs = require('fs');
+const readFile = denodeify(fs.readFile);
 
 module.exports = (app, meta, initPromises) => {
-
-	const actualAppListen = function () {
-		let serverPromise;
+	async function createServer(app) {
 		if (process.argv.indexOf('--https') > -1) {
-			const readFile = denodeify(fs.readFile);
-			serverPromise = Promise.all([
+			const [key, cert] = await Promise.all([
 				readFile(path.resolve(__dirname, '../../key.pem')),
 				readFile(path.resolve(__dirname, '../../cert.pem'))
 			])
-				.then(results => https.createServer({ key: results[0], cert: results[1] }, this));
+
+			return https.createServer({ key, cert }, app);
 		} else {
-			serverPromise = Promise.resolve(http.createServer(this));
+			return http.createServer(app);
 		}
+	}
+
+	async function actualAppListen(app, ...args) {
+		const server = await createServer(app)
 
-		return serverPromise.then(server => server.listen.apply(server, arguments));
+		return server.listen(...args);
 	};
 
-	app.listen = function () {
-		const args = [].slice.apply(arguments);
+	app.listen = async function (port, callback) {
+		// these middleware are attached in .listen so they're
+		// definitely after any middleware added by the app itself
 
 		// The error handler must be before any other error middleware
 		app.use(raven.errorHandler());
 
 		// Optional fallthrough error handler
-		app.use((err, req, res, next) => { //eslint-disable-line
+		app.use((err, req, res, next) => { // eslint-disable-line no-unused-vars
 			// The error id is attached to `res.sentry` to be returned
 			// and optionally displayed to the user for support.
 			res.statusCode = 500;
 			res.end(res.sentry + '\n');
 		});
 
-		const port = args[0];
-		const cb = args[1];
-		args[1] = function () {
+		function wrappedCallback() {
 			// HACK: Use warn so that it gets into Splunk logs
 			nLogger.warn({ event: 'EXPRESS_START', app: meta.name, port: port, nodeVersion: process.version });
-			return cb && cb.apply(this, arguments);
+
+			if(callback) {
+				return callback.apply(this, arguments);
+			}
 		};
 
-		return Promise.all(initPromises)
-			.then(() => {
-				metrics.count('express.start');
-				return actualAppListen.apply(app, args);
-			})
-			// Crash app if initPromises fail
-			.catch(err => setTimeout(() => {
+		try {
+			await Promise.all(initPromises)
+		} catch(err) {
+			// crash app if initPromises fail by throwing an error asynchronously outside of the promise
+			// TODO: better error handling
+			setTimeout(() => {
 				throw err;
-			}));
+			});
+		}
+
+		metrics.count('express.start');
```

Comment:
Would it be better if this line lived in `wrappedCallback` along side the similar log message? As that's more accurately when express has started.

I'm not sure what the benefit of having both is TBH.

Reply:
`express.start` isn't a log, it's for Graphite, our Grafana dashboards mark it on the time series. i deliberately haven't changed where this is happening in this PR because it's just a refactor. this metric has been counted in the exact same place [literally since 2015](https://github.com/Financial-Times/n-express/commit/a7612cc00d6bf9a17a898d166f66b5b6f4ec2b29), if we want to change it i think it should be more considered.

Reply:
Of course I know the difference 😄 . I get it's just a refactor. It's just that putting this on line 50 makes the two things that make the same event happen in the same place.

Reply:
if you think that's better, go ahead and raise a PR. it's been fine where it is for six years is all